### PR TITLE
Extract a Git_ref module

### PR DESCRIPTION
**Pull request by NathanReb**
This PR extracts a `Git_ref` module with an abstract type `t`. While doing this refactoring it appeared unclear what `tag` was used for exactly and having it passed around as a string doesn't help.

This is not perfect yet as I'm still trying to figure out the correct abstractions to use but at least it splits out a bit of `Opam_cmd`. I think it could be merged as is but I'll try to improve it a bit next week!

### CODE REVIEW

File: test/lib/test_git_ref.ml [hunk 1]
```diff
@@ -1,38 +1,39 @@
...
+      Alcotest.(check (option Testable.git_ref)) archive expected actual
     in
     (test_name, `Quick, test_fun)
   in
-  [ make_test ~archive:"git+http://a.com/user/repo" ~expected:(Some "master");
-    make_test ~archive:"git+https://a.com/user/repo" ~expected:(Some "master");
-    make_test ~archive:"git+https://a.com/user/repo#v1.2.3" ~expected:(Some "v1.2.3");
-    make_test ~archive:"git+https://github.com/user/repo#v1.2.3" ~expected:(Some "v1.2.3");
-    make_test ~archive:"git+ssh://a.com/user/repo#v1.2.3" ~expected:(Some "v1.2.3");
+  [ make_test ~archive:"git+http://a.com/user/repo" ~expected:(Some (make "master"));
```
Comment (Julow) [inline test/lib/test_git_ref.ml]: The test cases could be lighter:
```ocaml
let expect_ref archive ref = make_test ~archive ~expected:(Some (make ref))
let expect_none archive = make_test ~archive ~expected:None
```
```ocaml
expect_ref "git+http://a.com/user/repo" "master";
expect_none "git+file://a.com/user/repo/something"
```

File: lib/opam_cmd.ml [hunk 2]
```diff
@@ -107,22 +61,22 @@ let classify_package ~package ~dev_repo ~archive ~pins () =
...
           Logs.debug (fun l ->
-              l "Mapped %s -> %s" archive (match tag with None -> "??" | Some v -> v) );
+              l "Mapped %s -> %s" archive (match git_ref with None -> "??" | Some v -> Git_ref.to_string v) );
           match Uri.host uri with
           | Some "github.com" -> (
             match String.cuts ~empty:false ~sep:"/" (Uri.path uri) with
             | [ user; repo ] ->
-                let repo = strip_ext repo in
-                (`Github (user, repo), tag)
+                let repo = Duniverse_std.File.strip_ext repo in
```
Comment (Julow) [inline lib/opam_cmd.ml]: Github repos can contain dots and the `.git` extension is optional. This is a bug.

**Comment by NathanReb** [comment]
Yeah I'm actually not to happy with this myself. I think the local stdlib is okay but I'll get back to this kind of refactoring when #14 , #23  and #25 are in.

**Comment by NathanReb** [comment]
Closing this for now.

**Pull request closed without merging**